### DISCUSSION TITLE
Use ApexCharts in Nuxt3

### DISCUSSION THREAD

**Discussion by brunokunace** [nuxt/nuxt#16482] [Questions]
Hi all, i am try to running vue3-apexcharts in nuxt3 ssr.

i create the plugin with name `apexcharts.client.js` with this content:
```

import VueApexCharts from 'vue3-apexcharts'
export default defineNuxtPlugin(nuxtApp => {
    nuxtApp.vueApp.use(VueApexCharts)
})
```

the problem is when try to use <VueApexCharts /> in template, because this component not exist in first render.

the error is the common "window is undefined" or "unknown component VueApexCharts".

if i remove ssr from nuxt.config works fine.

in nuxt2, i use the <client-only> tag and works, but in nuxt3 maybe not have it anymore.

Can any help-me with this?

**Comment by benoitdemaegdt** [comment]
Hey @brunokunace  👋 

I guess your chart should not be rendered on server side.
You may have a look at [ClientOnly](https://v3.nuxtjs.org/guide/directory-structure/components#clientonly-component) tag 😃

**Comment by brunokunace** [comment]
hi @benoitdemaegdt , thanks for reply.

Yes, i am using it, but the plugin not initiate on client, even though i put .client. in plugin file name or put in nuxt.config the mode: client.

i receive the error:

ReferenceError: window is not defined
    at Object.<anonymous> (...node_modules/apexcharts/dist/apexcharts.common.js:6:360429)
    
why the plugin running in server and not in client?

**Reply by wobsoriano** [reply]
You still need to wrap it in a `<ClientOnly />` component

**Comment by wobsoriano** [comment]
https://github.com/nuxt/framework/discussions/3028#discussioncomment-2095655

**Comment by brunokunace** [comment]
thanks for answer @wobsoriano but, i have this inside a ClientOnly tag

<img width="1016" alt="image" src="https://user-images.githubusercontent.com/16597320/172926857-894684ad-4e03-471d-8edf-aeab20a9aeca.png">

**Reply by wobsoriano** [reply]
Try creating a custom component and wrap that component with `<ClientOnly />`

```html
<script setup lang="ts">
import VueApexCharts from 'vue3-apexcharts'

const chartOptions = {}
</script>

<template>
  <VueApexCharts :options="chartOptions" />
</template>
```

Then

```html
<script setup lang="ts">
import CustomApexChartComponent from '~/components/CustomApexChartComponent'
</script>

<template>
  <ClientOnly>
    <CustomApexChartComponent />
  </ClientOnly>
</template>
```

**Comment by brunokunace** [comment]
hi @wobsoriano, i try you solution, but have the same error, window of undefined, idk why nuxt import this in server if have a client only tag

**Reply by wobsoriano** [reply]
Create a simple repro with stackblitz then I can help.

**Reply by brunokunace** [reply]
its. most simple example https://stackblitz.com/edit/nuxt-starter-4wpumj?file=app.vue if remove comment in nuxt.config for disable ssr, thee example works

**Comment by wobsoriano** [comment]
You probably just missed something, but just creating a plugin out of it and using it works tho:

```ts
import VueApexCharts from 'vue3-apexcharts';

export default defineNuxtPlugin((nuxtApp) => {
  nuxtApp.vueApp.use(VueApexCharts);
});
```

```html
<script setup>
const chartOptions = {
  chart: {
    id: 'vuechart-example',
  },
  xaxis: {
    categories: [1991, 1992, 1993, 1994, 1995, 1996, 1997, 1998],
  },
};
const series = [
  {
    name: 'series-1',
    data: [30, 40, 35, 50, 49, 60, 70, 91],
  },
];
</script>

<template>
  <apexchart
    width="500"
    type="bar"
    :options="chartOptions"
    :series="series"
  ></apexchart>
</template>
```

https://stackblitz.com/edit/nuxt-starter-btzweo?file=plugins%2Fapexcharts.client.ts,app.vue

**Reply by zhouzilong2020** [reply]
This solves my problem!!! Thanks a lot!!!!

But I am still a little bit confused about what is going on here:
I use vue3-apextcharts, the wrapper package, in my code. I did create a new plugin file the same as yours, but it still complains ```window is not defined``` when I try to run ```Nuxt generate```.
Once I changed the usage from ```<vue-apex-charts />``` to ```<apexchat />```, this issue immediately disappeared. 

Wouldn't the plugin file has already set the mode of ```vue3-apexcharts``` to ```client```? Everything should be just right.

Any idea what happened?

**Reply by Bergrebell** [reply]
the stackblitz also throws the same error as i've mentioned in my comment below ... any ideas?

**Reply by matej-marjanovic** [reply]
Tried testing the StackBlitz example with Nuxt 3.rc12. Doesn't work if you load the page. However, if you load the page with `<apexchart>` commented out and then comment it back in, then HMR adds in the chart with no issues. Encapsulating it in `<ClientOnly>` appears to fix the issue.
StackBlitz: https://stackblitz.com/edit/nuxt-starter-gsvfln?file=app.vue

**Reply by wobsoriano** [reply]
Updated the example. Sorry! You should wrap your chart component within a `<ClientOnly>` component or use the `.client.vue` file extension :)

**Reply by jmarshall9120** [reply]
Jus did this, needed one more line of code.  Add this to the nuxt.config.ts
```
plugins: [{ src: "~/plugins/apexcharts.ts", mode: "client" }]
```

**Comment by Bergrebell** [comment]
with the solution provided it works but throws the following error:
```
[Vue warn]: Failed to resolve component: apexchart
If this is a native custom element, make sure to exclude it from component resolution via compilerOptions.isCustomElement.
[nuxt] [request error] Invalid value used as weak map key
```

**Reply by szulcus** [reply]
For me the problem was the lack of the "apexchart" package. After adding it, the chart started working

**Comment by flyingkaktus** [comment]
same issue. why is it so hard to create a proper documentation which packages are needed and so on? jeuss

**Reply by wobsoriano** [reply]
The demo above works - https://stackblitz.com/edit/nuxt-starter-btzweo?file=app.vue
